python_modules/dagster/dagster/_core/definitions/asset_layer.py

```diff
@@ -1,230 +1,205 @@
-from collections import defaultdict
 from collections.abc import Iterable, Mapping
-from typing import TYPE_CHECKING, AbstractSet, NamedTuple, Optional  # noqa: UP035
+from functools import cached_property
+from typing import TYPE_CHECKING, AbstractSet, Optional, Sequence, Union  # noqa: UP035
+
+from dagster_shared.record import record
 
 import dagster._check as check
 from dagster._core.definitions.asset_check_spec import AssetCheckKey
 from dagster._core.definitions.asset_key import AssetKey, EntityKey
 from dagster._core.definitions.dependency import NodeHandle, NodeInputHandle, NodeOutputHandle
 from dagster._core.definitions.graph_definition import GraphDefinition
-from dagster._core.errors import DagsterInvariantViolationError
 
 if TYPE_CHECKING:
     from dagster._core.definitions.asset_graph import AssetGraph, AssetNode
     from dagster._core.definitions.assets import AssetsDefinition
-    from dagster._core.definitions.partitions.mapping import PartitionMapping
 
 
-class AssetLayer(NamedTuple):
-    """Stores all of the asset-related information for a Dagster job. Maps each
-    input / output in the underlying graph to the asset it represents (if any), and records the
-    dependencies between each asset.
+@record(checked=False)
+class AssetLayerData:
+    """Data that relates asset-level information to a node in the execution graph."""
+
+    node_handle: NodeHandle
+    assets_def: "AssetsDefinition"
+
 
-    Args:
-        asset_key_by_node_input_handle (Mapping[NodeInputHandle, AssetOutputInfo], optional): A mapping
-            from a unique input in the underlying graph to the associated AssetKey that it loads from.
-        asset_keys_by_node_output_handle (Mapping[NodeOutputHandle, AssetOutputInfo], optional): A mapping
-            from a unique output in the underlying graph to the associated AssetOutputInfo.
+@record(checked=False)
+class AssetLayer:
+    """Stores all of the asset-related information for a Dagster job. Maps execution node data
+    to the backing AssetsDefinition, with a variety of convenience methods for common operations.
     """
 
+    data: Sequence[AssetLayerData]
     asset_graph: "AssetGraph"
-    assets_defs_by_node_handle: Mapping[NodeHandle, "AssetsDefinition"]
-    asset_keys_by_node_input_handle: Mapping[NodeInputHandle, AssetKey]
-    asset_keys_by_node_output_handle: Mapping[NodeOutputHandle, AssetKey]
-    check_key_by_node_output_handle: Mapping[NodeOutputHandle, AssetCheckKey]
-    node_output_handles_by_asset_check_key: Mapping[AssetCheckKey, NodeOutputHandle]
-    check_names_by_asset_key_by_node_handle: Mapping[
-        NodeHandle, Mapping[AssetKey, AbstractSet[str]]
-    ]
-    additional_asset_keys: AbstractSet[AssetKey]
-    outer_node_names_by_asset_key: Mapping[AssetKey, str] = {}
+    additional_asset_keys: AbstractSet[AssetKey] = set()
+    additional_asset_keys_by_input_handle: Mapping[NodeInputHandle, AssetKey] = {}
```

Comment:
im confused by these two that read like they should be in sync but appear inconsistently set

Reply:
updated!